File: TelerikUIExamples/TelerikUIExamples/SideDrawerTransitions.m
//
//  SideDrawerTransitions.m
//  TelerikUIExamples
//
//  Copyright (c) 2015 Telerik. All rights reserved.
//

#import "SideDrawerTransitions.h"
#import "SideDrawerHeaderView.h"

@implementation SideDrawerTransitions
{
    UIScrollView *_scrollView;
    CGFloat _buttonY;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _scrollView = [[UIScrollView alloc] initWithFrame:CGRectZero];
    _scrollView.scrollEnabled = YES;
    
    [self addButtonWithTitle:@"Push" target:self selector:@selector(pushTransition)];
    [self addButtonWithTitle:@"Reveal" target:self selector:@selector(revealTransition)];
    [self addButtonWithTitle:@"Reverse Slide Out" target:self selector:@selector(reverseSlideOutTransition)];
    [self addButtonWithTitle:@"Slide Along" target:self selector:@selector(slideAlongTransition)];
    [self addButtonWithTitle:@"Slide In On Top" target:self selector:@selector(slideInOnTopTransition)];
    [self addButtonWithTitle:@"Scale Up" target:self selector:@selector(scaleUpTransition)];
    [self addButtonWithTitle:@"Fade In" target:self selector:@selector(fadeInTransition)];
    [self addButtonWithTitle:@"Scale Down Pusher" target:self selector:@selector(scaleDownPusherTransition)];
    
    [self.sideDrawerView.mainView addSubview:_scrollView];
    self.sideDrawerView.backgroundColor = [UIColor grayColor];
    
    self.sideDrawer.style.shadowMode = TKSideDrawerShadowModeSideDrawer;
}

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    _scrollView.frame = CGRectMake(0, 44, CGRectGetWidth(self.view.bounds), CGRectGetHeight(self.sideDrawerView.mainView.bounds) - 44);
}

- (void)pushTransition
{
    TKSideDrawer *sideDrawer = self.sideDrawer;
    sideDrawer.fill = [TKSolidFill solidFillWithColor:[UIColor grayColor]];
    sideDrawer.transition = TKSideDrawerTransitionTypePush;
    // >> drawer-transitions-duration
    sideDrawer.transitionDuration = 0.5;
    // << drawer-transitions-duration
    sideDrawer.headerView = [[SideDrawerHeaderView alloc] initWithButton:NO target:nil selector:nil];
    [self showSideDrawer];
}

- (void)revealTransition
{
    TKSideDrawer *sideDrawer = self.sideDrawer;
    sideDrawer.fill = [TKSolidFill solidFillWithColor:[UIColor grayColor]];
    
    // >> drawer-transition
    sideDrawer.transition = TKSideDrawerTransitionTypeReveal;
    // << drawer-transition
    
    sideDrawer.headerView = [[SideDrawerHeaderView alloc] initWithButton:NO target:nil selector:nil];
    [self showSideDrawer];
}

- (void)reverseSlideOutTransition
{
    TKSideDrawer *sideDrawer = self.sideDrawer;
    sideDrawer.fill = [TKSolidFill solidFillWithColor:[UIColor grayColor]];
    sideDrawer.transition = TKSideDrawerTransitionTypeReverseSlideOut;
    sideDrawer.headerView = [[SideDrawerHeaderView alloc] initWithButton:NO target:nil selector:nil];
    [self showSideDrawer];
}

- (void)slideAlongTransition
{
    TKSideDrawer *sideDrawer = self.sideDrawer;
    sideDrawer.fill = [TKSolidFill solidFillWithColor:[UIColor grayColor]];
    sideDrawer.transition = TKSideDrawerTransitionTypeSlideAlong;
    sideDrawer.headerView = [[SideDrawerHeaderView alloc] initWithButton:NO target:nil selector:nil];
    [self showSideDrawer];
}

- (void)slideInOnTopTransition
{
    TKSideDrawer *sideDrawer = self.sideDrawer;
    sideDrawer.fill = [TKSolidFill solidFillWithColor:[UIColor clearColor]];
    sideDrawer.transition = TKSideDrawerTransitionTypeSlideInOnTop;
    sideDrawer.headerView = [[SideDrawerHeaderView alloc] initWithButton:YES target:self selector:@selector(dismissSideDrawer)];
    [self showSideDrawer];
}

- (void)scaleUpTransition
{
    TKSideDrawer *sideDrawer = self.sideDrawer;
    sideDrawer.fill = [TKSolidFill solidFillWithColor:[UIColor grayColor]];
    sideDrawer.transition = TKSideDrawerTransitionTypeScaleUp;
    sideDrawer.headerView = [[SideDrawerHeaderView alloc] initWithButton:NO target:nil selector:nil];
    [self showSideDrawer];
}

- (void)fadeInTransition
{
    TKSideDrawer *sideDrawer = self.sideDrawer;
    sideDrawer.fill = [TKSolidFill solidFillWithColor:[UIColor grayColor]];
    sideDrawer.transition = TKSideDrawerTransitionTypeFadeIn;
    sideDrawer.headerView = [[SideDrawerHeaderView alloc] initWithButton:YES target:self selector:@selector(dismissSideDrawer)];
    [self showSideDrawer];
}

- (void)scaleDownPusherTransition
{
    TKSideDrawer *sideDrawer = self.sideDrawer;
    sideDrawer.fill = [TKSolidFill solidFillWithColor:[UIColor grayColor]];
    sideDrawer.transition = TKSideDrawerTransitionTypeScaleDownPusher;
    sideDrawer.headerView = [[SideDrawerHeaderView alloc] initWithButton:NO target:self selector:@selector(dismissSideDrawer)];
    [self showSideDrawer];
}

- (void)addButtonWithTitle:(NSString *)title target:(id)target selector:(SEL)selector
{
    CGSize titleSize = [title sizeWithAttributes:@{ NSFontAttributeName : [UIFont systemFontOfSize:18] }];
    UIButton *button = [[UIButton alloc] initWithFrame:CGRectMake(15, 15 + _buttonY, titleSize.width, 44)];
    button.titleLabel.font = [UIFont systemFontOfSize:14];
    [button.layer setBorderWidth:1.0];
    [button.layer setBorderColor:[UIColor whiteColor].CGColor];
    [button.layer setCornerRadius:3.f];
    [button setTitle:title forState:UIControlStateNormal];
    [button setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [button addTarget:target action:selector forControlEvents:UIControlEventTouchUpInside];
    [_scrollView addSubview:button];
    _buttonY += 50;
    _scrollView.contentSize = CGSizeMake(fmax(CGRectGetWidth(button.frame), _scrollView.contentSize.width), _buttonY + 15 + self.view.bounds.origin.y);
}

#pragma mark TKSideDrawerDelegate

- (void)sideDrawer:(TKSideDrawer *)sideDrawer updateVisualsForItem:(NSInteger)item inSection:(NSInteger)section
{
    TKSideDrawerItem *currentItem = ((TKSideDrawerSection *)sideDrawer.sections[section]).items[item];
    currentItem.style.contentInsets = UIEdgeInsetsMake(0, -5, 0, 0);
    currentItem.style.separatorColor = [TKSolidFill solidFillWithColor:[UIColor clearColor]];
}

- (void)sideDrawer:(TKSideDrawer *)sideDrawer updateVisualsForSection:(NSInteger)sectionIndex
{
    TKSideDrawerSection *section = sideDrawer.sections[sectionIndex];
    section.style.contentInsets = UIEdgeInsetsMake(0, -15, 0, 0);
}

@end

File: TelerikUIExamplesInSwift/TelerikUIExamplesInSwift/SideDrawerTransitions.swift
//
//  SideDrawerTransitions.swift
//  TelerikUIExamplesInSwift
//
//  Copyright (c) 2015 Telerik. All rights reserved.
//

class SideDrawerTransitions: SideDrawerGettingStarted {
    
    var scrollView:UIScrollView = UIScrollView()
    var buttonY:CGFloat = 0
    var v = UIView()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        scrollView = UIScrollView(frame: CGRectZero)
        scrollView.scrollEnabled = true
        
        self.addButtonWithTitle("Push", target: self, selector: #selector(SideDrawerTransitions.pushTransition))
        self.addButtonWithTitle("Reveal", target: self, selector: #selector(SideDrawerTransitions.revealTransition))
        self.addButtonWithTitle("Reverse Slide Out", target: self, selector: #selector(SideDrawerTransitions.reverseSlideOutTransition))
        self.addButtonWithTitle("Slide Along", target: self, selector: #selector(SideDrawerTransitions.slideAlongTransition))
        self.addButtonWithTitle("Slide In On Top", target: self, selector: #selector(SideDrawerTransitions.slideInOnTopTransition))
        self.addButtonWithTitle("Scale Up", target: self, selector: #selector(SideDrawerTransitions.scaleUpTransition))
        self.addButtonWithTitle("Fade In", target: self, selector: #selector(SideDrawerTransitions.fadeInTransition))
        self.addButtonWithTitle("Scale Down Pusher", target: self, selector: #selector(SideDrawerTransitions.scaleDownPusherTransition))
        
        self.sideDrawerView.mainView.addSubview(scrollView)
        self.sideDrawerView.backgroundColor = UIColor.grayColor()
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        scrollView.frame = CGRectMake(0, 44, CGRectGetWidth(self.sideDrawerView.bounds), CGRectGetHeight(self.sideDrawerView.bounds) - 44)
    }
    
    func addButtonWithTitle(title: String, target: AnyObject, selector: Selector) {
        let titleString: NSString = title as NSString
        let titleSize = titleString.sizeWithAttributes([NSFontAttributeName : UIFont.systemFontOfSize(18.0)])
        let button = UIButton(frame: CGRectMake(15, 15 + self.buttonY, titleSize.width, 44))
        button.titleLabel?.font = UIFont.systemFontOfSize(14)
        button.layer.borderWidth = 1.0
        button.layer.borderColor = UIColor.whiteColor().CGColor
        button.layer.cornerRadius = 3.0
        button.setTitle(title, forState: UIControlState.Normal)
        button.setTitleColor(UIColor.whiteColor(), forState: UIControlState.Normal)
        button.addTarget(self, action: selector, forControlEvents: UIControlEvents.TouchUpInside)
        scrollView.addSubview(button)
        self.buttonY += 50
        self.scrollView.contentSize = CGSizeMake(fmax(CGRectGetWidth(button.frame), self.scrollView.contentSize.width), self.buttonY + 15)
    }
    
    func pushTransition() {
        let sideDrawer = self.sideDrawerView.sideDrawers[0]
        sideDrawer.transition = TKSideDrawerTransitionType.Push
        sideDrawer.fill = TKSolidFill(color: UIColor.grayColor())
        sideDrawer.headerView = SideDrawerHeaderView(addButton: false, target: nil, selector: nil)
        self.showSideDrawer()
    }
    
    func revealTransition() {
        let sideDrawer = self.sideDrawerView.sideDrawers[0]
        sideDrawer.transition = TKSideDrawerTransitionType.Reveal
        sideDrawer.fill = TKSolidFill(color: UIColor.grayColor())
        sideDrawer.headerView = SideDrawerHeaderView(addButton: false, target: nil, selector: nil)
        self.showSideDrawer()
    }
    
    func reverseSlideOutTransition() {
        let sideDrawer = self.sideDrawerView.sideDrawers[0]
        sideDrawer.transition = TKSideDrawerTransitionType.ReverseSlideOut
        sideDrawer.fill = TKSolidFill(color: UIColor.grayColor())
        sideDrawer.headerView = SideDrawerHeaderView(addButton: false, target: nil, selector: nil)
        self.showSideDrawer()
    }
    
    func slideAlongTransition() {
        let sideDrawer = self.sideDrawerView.sideDrawers[0]
        sideDrawer.transition = TKSideDrawerTransitionType.SlideAlong
        sideDrawer.fill = TKSolidFill(color: UIColor.grayColor())
        sideDrawer.headerView = SideDrawerHeaderView(addButton: false, target: nil, selector: nil)
        self.showSideDrawer()
    }
    
    func slideInOnTopTransition() {
        let sideDrawer = self.sideDrawerView.sideDrawers[0]
        sideDrawer.transition = TKSideDrawerTransitionType.SlideInOnTop
        sideDrawer.fill = TKSolidFill(color: UIColor.clearColor())
        sideDrawer.headerView = SideDrawerHeaderView(addButton: true, target: self, selector: #selector(SideDrawerTransitions.dismissSideDrawer))
        self.showSideDrawer()
    }
    
    func scaleUpTransition() {
        let sideDrawer = self.sideDrawerView.sideDrawers[0]
        sideDrawer.transition = TKSideDrawerTransitionType.ScaleUp
        sideDrawer.fill = TKSolidFill(color: UIColor.grayColor())
        sideDrawer.headerView = SideDrawerHeaderView(addButton: false, target: nil, selector: nil)
        self.showSideDrawer()
    }
    
    func fadeInTransition() {
        let sideDrawer = self.sideDrawerView.sideDrawers[0]
        sideDrawer.transition = TKSideDrawerTransitionType.FadeIn
        sideDrawer.fill = TKSolidFill(color: UIColor.grayColor())
        sideDrawer.headerView = SideDrawerHeaderView(addButton: true, target: self, selector: #selector(SideDrawerTransitions.dismissSideDrawer))
        self.showSideDrawer()
    }
    
    func scaleDownPusherTransition() {
        let sideDrawer = self.sideDrawerView.sideDrawers[0]
        sideDrawer.transition = TKSideDrawerTransitionType.ScaleDownPusher
        sideDrawer.fill = TKSolidFill(color: UIColor.grayColor())
        sideDrawer.headerView = SideDrawerHeaderView(addButton: false, target: self, selector: #selector(SideDrawerTransitions.dismissSideDrawer))
        self.showSideDrawer()
    }
}
